File: Data Structures & Algorithms/rotate-array/submission-1.java
class Solution {
    private static void rev(int[] arr, int i, int j) {
        while(i < j) {
            int tt = arr[i];
            arr[i] = arr[j];
            arr[j] = tt;
            i++;
            j--;
        }
    }
    public void rotate(int[] nums, int k) {
        int L = nums.length;
        k%=L;
        rev(nums, 0, L - 1);
        rev(nums, 0, k - 1);
        rev(nums, k, L - 1);

        // for(int i=0; i < L; i++) {
        //     int n = ((i+k) % L);
        //     int tt = nums[i];
        //     nums[i] = nums[n];
        //     nums[n] = tt;
        // }
    }
}

File: Data Structures & Algorithms/subsets-ii/submission-1.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        def solve(idx, res, buff):
            if idx == len(nums):
                # res.append(buff[:])
                res.append(buff[:])
                return
            # pick
            buff.append(nums[idx])
            solve(idx + 1, res, buff)
            buff.pop()
            
            # skip duplicates for this nums[idx]
            while idx + 1 < len(nums) and nums[idx] == nums[idx + 1]:
                idx += 1
            # not pick
            solve(idx + 1, res, buff)

        res, buff, idx = [], [], 0

        nums.sort()

        solve(idx, res, buff)

        return res
        